### PR TITLE
Some initial thoughts and suggestions

### PR DESCRIPTION
**Pull request by jackrusher**

### CODE REVIEW

File: a.el [hunk 1]
```diff
@@ -23,8 +23,11 @@
 
 ;;; Code:
 
-(require 'dash)
+(eval-when-compile (require 'subr-x))
 
+(require 'dash)
+;; if you're going to use dash, you might also want ht:
+;; https://github.com/Wilfred/ht.el/blob/master/ht.el
```
Comment (plexus) [inline a.el:30]: I'm not against using it, just didn't feel the need yet.

File: a.el [hunk 2]
```diff
@@ -74,6 +77,7 @@
               copy)
           (vconcat coll (-repeat (- k (length coll)) nil) (list v)))))
 
+   ;; XXX vectorp duplicated? placeholder for hash-table, maybe?
```
Comment (plexus) [inline a.el:80]: yeah, that's fixed on master already.

File: a.el [hunk 3]
```diff
@@ -84,6 +88,8 @@
 
 (defun a-assoc (coll &rest kvs)
   "Return an updated collection, associating values with keys."
+  (when (not (evenp (a-count kvs)))
+    (user-error "a-assoc requires an even number of arguments!"))
```
Comment (plexus) [inline a.el:92]: 👍

File: a.el [hunk 4]
```diff
@@ -124,10 +126,12 @@
     (length coll))
 
    ((hash-table-p coll)
-    (length (a-keys coll)))))
+    (hash-table-count coll))))
 
+;; 1) is it ok that collections of different types can be equal?
```
Comment (plexus) [inline a.el:131]: Yes, that's by design, the idea being that their "value" is the same (the way a Clojure seq and vector can be equal).

Thinking about it though we could have a separate `a-eql` which checks type, and `a-equal` which is more lenient.

File: a.el [hunk 5]
```diff
@@ -124,10 +126,12 @@
...
+;; 2) early termination would be nice for larger collections
```
Comment (plexus) [inline a.el:132]: That's true. The `(and bool ,,,)` short circuits so you skip any remaining value lookups and equality checks, but it'll still go through the rest of the loop.

In general I haven't been too concerned with performance (yet), assuming that if you're doing this style of programming performance isn't your main concern either, but performance improvements are of course very welcome.

File: a.el [hunk 6]
```diff
@@ -124,10 +126,12 @@
...
 (defun a-equal (a b)
-  "Reduce an associative collection, starting with an initial value of FROM. The reducing functions receives the intermediate value, key, and value."
+  "Compare collections a and b for equality. Return true or false."
```
Comment (plexus) [inline a.el:134]: Thanks, good catch! Probably should mention that it compares the set of key value pairs regardless of type. This one is also a bit of an odd one out in that it doesn't deal with vectors.

File: a.el [hunk 7]
```diff
@@ -143,10 +147,15 @@
...
 
+;; TODO a-merge-with 
+
 (defun a-alist (&rest kvs)
   "Create an association list from the given keys and values, provided as a single list of arguments. e.g. (a-alist :foo 123 :bar 456)"
   (mapcar (lambda (kv) (cons (car kv) (cadr kv))) (-partition 2 kvs)))
 
+;; XXX shouldn't this fn use the top of the top level collection to
+;; create children? that is, shouldn't (a-assoc-in (make-hash-table)
+;; ...) produce nested hash tables?
```
Comment (plexus) [inline a.el:158]: Not a terribly strong opinion on this one. I can see the reasoning, but I wouldn't say that's particularly intuitive behavior either. I'm fine with this function (and the lib in general) leaning a bit towards preferring alists.

**Comment by plexus** [comment]
Thanks a lot, Jack! Really appreciated.

I left a bunch of comments. Let me know if you want to push anything more to this branch, otherwise I'm fine with merging it and dealing with your comments on master.

File: a.el (hunk 7 above)
Comment (jackrusher) [reply a.el:158]: Hm. I feel like a large part of the value this library is providing a uniform interface over the different associative containers in elisp. If that's true, then preserving the type of a container manipulated through that interface would be quite important.

**Comment by jackrusher** [comment]
Merge whenever you like! I'll circle around in a week or so and have another look. :)

**Pull request merged**